cpp/src/arrow/buffer.h

```diff
@@ -137,6 +140,25 @@ class ARROW_EXPORT Buffer {
   /// \return a new Buffer instance
   static std::shared_ptr<Buffer> FromString(std::string data);
 
+  /// \brief Construct an immutable buffer that takes ownership of the contents
+  /// of an std::vector (without copying it).
+  ///
+  /// \param[in] vec a vector to own
+  /// \return a new Buffer instance
+  template <typename T>
+  static std::shared_ptr<Buffer> FromVectorOfTrivial(std::vector<T> vec) {
+    static_assert(std::is_trivial_v<T>);
```

Comment:
Instead of the wacky name, why not make the compile error more explicit?
```suggestion
  static std::shared_ptr<Buffer> FromVector(std::vector<T> vec) {
    static_assert(std::is_trivial_v<T>, "Buffer::FromVector needs a trivial type");
```

Reply:
I was following @felipecrv's comment. This seems fine, though; the compilation error should remove any doubts about what the function can convert

Reply:
https://github.com/apache/arrow/pull/36418#discussion_r1248323736